canonical-websites/insights.ubuntu.com
# Cache in memory rather than disk

We have been caching requests to a cache.sqlite file on disk. This means that the cache will persist if the process restarts, but it is slower.

In Kubernetes, there's not really a whole lot of point persisting the cache between different processes, because a single container runs a single process, and usually if the process crashes the container is
recreated.

Caching in memory is about 0.05 seconds quicker in local testing (curl responses are around 0.20s rather than 0.25s). This is not insignificant.

However, in local development, the disk cache is genuinely useful, as the process will be restarted a lot, and we don't want to have to put up with 10 second page loads all the time. For that reason I use REQUEST_CACHE_BACKEND=sqlite in local development.

QA
--

First test local development:

``` bash
rm cache.sqlite  # Delete the sqlite cache
./run
```

In a separate window, time a few requests:

``` bash
$ time curl -s http://0.0.0.0:8023/ > /dev/null  # The first request should be slow
real    0m11.004s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.231s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.218s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.237s
$ ls cache.sqlite  # A cache.sqlite file should have been created
cache.sqlite
```

Now edit `.env` and comment out the `REQUEST_CACHE_BACKEND` variable.

Now start the server again, and delete the cache:

``` bash
rm cache.sqlite  # Delete the sqlite cache
./run
```

And in a separate window, again run a few requests. The cache should respond faster:

``` bash
$ time curl -s http://0.0.0.0:8023/ > /dev/null  # The first request should be slow
real    0m11.004s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.191s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.181s
$ time curl -s http://0.0.0.0:8023/ > /dev/null
real    0m0.189s
$ ls cache.sqlite  # A cache.sqlite file should *not* have been created
ls: cannot access 'cache.sqlite': No such file or directory
```
